lib/anno_utils.pm

```diff
@@ -0,0 +1,3570 @@
+
+    my %aa_stats = $self->_generate_stats_from_aa($aa_ref);
+    my %aa_gff_stats = $self->_generate_stats_from_gffContents($aa_gff_conts);
+    my %subsys_info = $self->_fetch_subsystem_info();
+
+    my @html_files = $self->_write_html_from_stats(\%aa_stats,
+                                                   \%aa_gff_stats,
+                                                   \%subsys_info,
+                                                   $func_tab);
+
+    my $kbr = new installed_clients::KBaseReportClient($self->{call_back_url});
+    my $report_message = $msg;
+    my $report_info = $kbr->create_extended_report(
+            {"message"=>$report_message,
+             "objects_created"=>[{"ref"=>$aa_ref, "description"=>"RAST re-annotated genome"}],
+             "html_links"=> \@html_files,
+             "direct_html_link_index"=> 0,
+             "html_window_height"=> 366,
+             "report_object_name"=>"kb_RAST_genome_report_".$self->_create_uuid(),
+             "workspace_name"=>$gn_info->[7]
+        });
+
+    return {"output_genome_ref"=>$aa_ref,
+            "workspace_name"=>$report_info->{workspace_name},
+            "report_name"=>$report_info->{name},
+            "report_ref"=>$report_info->{ref}};
+}
+
+#
+## Check the given ref string to make sure it conforms to the expected
+## object reference format and actually exists in KBase.
+## Return the object data if it passes, die otherwise.
+#
+sub _fetch_object_data {
+    my ($self, $obj_ref) = @_;
+    my $ret_obj_data = {};
+
+    unless ($obj_ref =~ m/^\d+\/\d+\/\d+$/) {
+		print "invalid object reference: $obj_ref\n";
+        return $ret_obj_data;
+    }
+
+    eval {
+        $ret_obj_data = $self->{ws_client}->get_objects2(
+                            {'objects'=>[{ref=>$obj_ref}]}
+                        )->{data}->[0]->{data};
+    };
+    if ($@) {
+        croak "ERROR Workspace.get_objects2 failed to access $obj_ref:$@\n";
+    }
+    return $ret_obj_data;
+}
+
+
+#
+## Check the given ref string to make sure it conforms to the expected
+## object reference format and actually exists in KBase.
+## Return the object info if it passes, undef otherwise.
+#
+sub _fetch_object_info {
+    my ($self, $obj_ref) = @_;
+
+    my $ret_obj_info = undef;
+    eval {
+        $ret_obj_info = $self->{ws_client}->get_object_info3(
+                                 {objects=>[{ref=>$obj_ref}]}
+                        )->{infos}->[0];
+    };
+    if ($@) {
+        print "INFO: Workspace.get_object_info3 failed to access $obj_ref.\n";
+        #print "ERROR message:$@\n";
+	    return undef;
+    }
+    else {
+        print "INFO: object info for $obj_ref------\n".Dumper($ret_obj_info);
+        return $ret_obj_info;
+    }
+}
+
+# create a 12 char string unique enough here
+sub _create_uuid {
+    my $self = shift;
+
+    my $str_uuid = qx(uuidgen);
+    chomp $str_uuid;
+    $str_uuid = substr($str_uuid, -12);
+    return $str_uuid;
+}
```

Comment:
isn't this code repeated in another file? Don't maintain two copies of identical code, extract it out into another module instead.